sdk/opendp/whitenoise/synthesizers/dpgan/dpgan.py

```diff
@@ -0,0 +1,184 @@
+            ed = st + item[0]
+            data_t.append(torch.tanh(data[:, st:ed]))
+            st = ed
+        elif item[1] == 'softmax':
+            ed = st + item[0]
+            data_t.append(F.gumbel_softmax(data[:, st:ed], tau=0.2))
+            st = ed
+        else:
+            continue
+    return torch.cat(data_t, dim=1)
+
+class DPGANSynthesizer(SDGYMBaseSynthesizer):
+    def __init__(self,
+                 latent_dim=128,
+                 gen_dim=(256, 256),
+                 dis_dim=(256, 256),
+                 batch_size=256,
+                 epochs=1000,
+                 n_critics=5, 
+                 delta=1e-5,
+                 budget=3.0):
+        self.latent_dim = latent_dim
+        self.gen_dim = gen_dim
+        self.dis_dim = dis_dim
+        self.batch_size = batch_size
+        self.epochs = epochs
+        self.n_critics = n_critics
+        self.delta = delta
+        self.budget = budget
+        
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    
+    def fit(self, data, categorical_columns=tuple(), ordinal_columns=tuple()):
+        self.transformer = GeneralTransformer()
+        self.transformer.fit(data, categorical_columns, ordinal_columns)
+        data = self.transformer.transform(data)
+        dataset = TensorDataset(torch.from_numpy(data.astype('float32')).to(self.device))
+        dataloader = DataLoader(dataset, batch_size=self.batch_size, shuffle=True, drop_last=True)
+        
+        data_dim = self.transformer.output_dim
+        
+        self.generator = Generator(self.latent_dim, self.gen_dim, data_dim)
+        self.generator = utils.convert_batchnorm_modules(self.generator).to(self.device)
+        
+        discriminator = Discriminator(data_dim, self.dis_dim)
+        discriminator = utils.convert_batchnorm_modules(discriminator).to(self.device)
+        
+        optimizer_d = optim.RMSprop(discriminator.parameters(), lr=4e-4)
+        
+        privacy_engine = PrivacyEngine(
+            discriminator,
+            batch_size=self.batch_size,
+            sample_size=len(data),
+            alphas=[1 + x / 10.0 for x in range(1, 100)] + list(range(12, 64)),
+            noise_multiplier=5.0,
+            max_grad_norm=1.0,
+            clip_per_layer=True
+        )
+        
+        privacy_engine.attach(optimizer_d)
+        optimizer_g = optim.RMSprop(self.generator.parameters(), lr=1e-4)
+        
+        for epoch in range(self.epochs):
+            for i, data in enumerate(dataloader):
+                discriminator.zero_grad()
+                
+                real_data = data[0].to(self.device)
+                
+                # train with fake data
+                noise = torch.randn(self.batch_size, self.latent_dim, 1, 1)
+                noise = noise.view(-1, self.latent_dim)
+                fake_data = self.generator(noise)
+                fake_data = apply_activate(fake_data, self.transformer.output_info)
```

Comment:
The issue becomes here, where we want to attach a different activation based on the datatype: softmax for categorical, tanh for continuous